### ISSUE TITLE
Inquiry about normalization discrepancy in Discriminator (train vs. eval)

### ISSUE THREAD

**Issue by TerranLord** [showlab/Adv-GRPO#4]
Hi,

Thanks for the excellent work! I have a quick question regarding the implementation of the Discriminator.

I noticed that the DINOv2 tokens are normalized before the classification head during [inference](https://github.com/showlab/Adv-GRPO/blob/main/adv_grpo/rewards.py#L411), but not during [training](https://github.com/showlab/Adv-GRPO/blob/main/scripts/train_sd3_fast_dino_patch.py#L190). I was wondering if this inconsistency is intentional or if it might be a bug.

Looking forward to your reply. Best regards.

**Comment by maobenz** [comment]
Hi,

Thanks a lot for taking a close look at the implementation and for pointing this out.

You are right that there is a difference between training and inference. During training, we directly feed the raw DINOv2 CLS tokens into the head:

`logits_real_cls = head(cls_real)`

while during inference (reward computation), we normalize both CLS and patch embeddings before applying the same head:

```
cls_emb = cls_emb / (cls_emb.norm(dim=-1, keepdim=True) + 1e-6)
sampled_patches = sampled_patches / (sampled_patches.norm(dim=-1, keepdim=True) + 1e-6)
```

This was not a strictly intentional design choice, but more of a practical detail when reusing the DINO features for reward computation.

In practice, this difference does not have a significant impact. The main reason is that the head is a simple linear layer, which can absorb feature scaling during training. Also, DINOv2 features tend to have relatively stable magnitudes, so the mismatch between normalized and unnormalized features is limited.

From another perspective, the normalization at inference makes the scoring closer to a cosine-similarity-like behavior, which we found to be reasonably stable when aggregating patch-level scores.

That said, we agree that making training and inference fully consistent (e.g., normalizing features in both cases) would be a cleaner design, and is definitely a reasonable alternative.

Thanks again for raising this!

Best regards,
